core/store/src/trie/state_parts.rs

```diff
@@ -341,50 +336,49 @@ impl Trie {
     fn find_child_in_dfs_order(
         &self,
         memory_threshold: u64,
-        node: &mut TrieNodeWithSize,
+        node: &mut TrieStorageNodeWithSize,
         memory_skipped: &mut u64,
         key_nibbles: &mut Vec<u8>,
     ) -> Result<bool, StorageError> {
         *memory_skipped += node.node.memory_usage_direct();
 
         match &node.node {
-            TrieNode::Leaf(key, _) => {
-                let (slice, _) = NibbleSlice::from_encoded(key);
+            TrieStorageNode::Empty => Ok(false),
+            TrieStorageNode::Leaf { extension, .. } => {
+                let (slice, _) = NibbleSlice::from_encoded(extension);
                 key_nibbles.extend(slice.iter());
 
                 // Leaf must contain value, so we found the boundary.
                 Ok(false)
             }
-            TrieNode::Branch(children, value_handle) => {
-                if *memory_skipped > memory_threshold && value_handle.is_some() {
+            TrieStorageNode::Branch { children, value } => {
+                if *memory_skipped > memory_threshold && value.is_some() {
                     // If we skipped enough memory and found some value, we found the boundary.
                     return Ok(false);
                 }
 
-                let mut iter = children.iter();
-                while let Some((index, child)) = iter.next() {
-                    let NodeHandle::Hash(h) = child;
-                    let child = self.retrieve_node(h)?.unwrap().1;
-                    if *memory_skipped + child.memory_usage > memory_threshold {
-                        core::mem::drop(iter);
-                        key_nibbles.push(index);
-                        *node = child;
-                        return Ok(true);
+                for index in 0..16 {
+                    if let Some(child) = &children[index] {
```

Comment:
Might be a good idea to implement an iter on top of generic children like we had for raw trie node!